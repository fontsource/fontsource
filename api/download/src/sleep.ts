// Have a sleep timer that kills the worker after 2 minutes
export const SLEEP_MINUTES = 2;

let sleepTimeout: NodeJS.Timeout;

export const keepAwake = (minutes: number) => {
	if (sleepTimeout) {
		clearTimeout(sleepTimeout);
	}

	// eslint-disable-next-line unicorn/no-process-exit
	sleepTimeout = setTimeout(() => process.exit(0), 1000 * 60 * minutes);
};
