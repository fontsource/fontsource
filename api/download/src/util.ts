import { type IDResponse } from 'common-api/types';
import { splitTag } from 'common-api/util';
import { StatusError } from 'itty-router';

import { type Manifest } from './types';

type ManifestGenerator = Omit<
	IDResponse,
	| 'variants'
	| 'defSubset'
	| 'license'
	| 'type'
	| 'family'
	| 'lastModified'
	| 'category'
>;

export const generateManifestItem = async (
	tag: string,
	file: string,
	metadata: ManifestGenerator,
): Promise<Manifest> => {
	const { id, version } = await splitTag(tag);
	let [filename, extension] = file.split('.');
	const [subset, weight, style] = filename.split('-');
	if (!subset || !weight || !style) {
		throw new StatusError(400, 'Bad Request. Invalid filename.');
	}

	if (id !== metadata.id) {
		throw new StatusError(400, 'Bad Request. Invalid ID.');
	}

	// If the extension is ttf, change it to woff since jsdelivr doesn't store tff
	// and we convert it from woff to ttf on the fly
	if (extension === 'ttf') {
		extension = 'woff';
	}

	return {
		id,
		subset,
		weight: Number(weight),
		style,
		variable: metadata.variable,
		extension,
		version,
		url: `https://cdn.jsdelivr.net/npm/@fontsource/${id}@${version}/files/${id}-${file.replace(
			'.ttf',
			'.woff',
		)}`,
	};
};

export const generateManifest = async (
	tag: string,
	metadata: ManifestGenerator,
): Promise<Manifest[]> => {
	const { id, version } = await splitTag(tag);

	if (id !== metadata.id) {
		throw new StatusError(400, 'Bad Request. Invalid font ID.');
	}

	// Generate manifest
	const manifest: Manifest[] = [];

	// Unicode range is needed to include numbered subset files, but it may include duplicates
	const subsetKeys = [
		...new Set([
			...Object.keys(metadata.unicodeRange).map(
				(key) => key.replace('[', '').replace(']', ''), // Files should be stored without brackets
			),
			...metadata.subsets,
		]),
	];

	for (const subset of subsetKeys) {
		for (const weight of metadata.weights) {
			for (const style of metadata.styles) {
				for (const extension of ['woff2', 'woff']) {
					manifest.push({
						id,
						subset,
						weight,
						style,
						variable: metadata.variable,
						extension,
						version,
						// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
						url: `https://cdn.jsdelivr.net/npm/@fontsource/${metadata.id}@${version}/files/${metadata.id}-${subset}-${weight}-${style}.${extension}`,
					});
				}
			}
		}
	}

	return manifest;
};
