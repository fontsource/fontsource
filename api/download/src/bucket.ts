import { info } from 'diary';
import { StatusError } from 'itty-router';

import type { Manifest, ManifestVariable } from './manifest';
import { keepAwake, SLEEP_MINUTES } from './sleep';

type R2Object = string;

interface ListBucket {
	status: number;
	objects: R2Object[];
}

interface R2UploadedPart {
	etag: string;
	partNumber: number;
}

type BucketPath = Pick<
	Manifest,
	'id' | 'subset' | 'weight' | 'style' | 'extension' | 'version'
>;

type BucketPathVariable = Pick<
	ManifestVariable,
	'id' | 'subset' | 'axes' | 'style' | 'version'
>;

export const bucketPath = ({
	id,
	subset,
	weight,
	style,
	extension,
	version,
}: BucketPath) => `${id}@${version}/${subset}-${weight}-${style}.${extension}`;

export const bucketPathVariable = ({
	id,
	subset,
	axes,
	style,
	version,
}: BucketPathVariable) =>
	`${id}@${version}/variable/${subset}-${axes}-${style}.woff2`;

const handleBucketError = (resp: Response, msg: string) => {
	if (resp.status === 401) {
		throw new StatusError(
			401,
			'Unauthorized. Please check your UPLOAD_KEY environment variable.',
		);
	}

	throw new StatusError(500, `Internal Server Error. ${msg}`);
};

export const listBucket = async (prefix: string) => {
	keepAwake(SLEEP_MINUTES);

	const resp = await fetch(`https://upload.fontsource.org/list/${prefix}`, {
		headers: {
			Authorization: `Bearer ${
				// biome-ignore lint/style/noNonNullAssertion: <explanation>
				process.env.UPLOAD_KEY!
			}`,
		},
	});
	if (!resp.ok) {
		handleBucketError(resp, 'Unable to list bucket.');
	}

	return await resp.json<ListBucket>();
};

const abortMultiPartUpload = async (
	bucketPath: string,
	uploadId: string,
	msg?: string,
) => {
	const resp = await fetch(
		`https://upload.fontsource.org/multipart/${bucketPath}?action=mpu-abort`,
		{
			method: 'DELETE',
			headers: {
				Authorization: `Bearer ${
					// biome-ignore lint/style/noNonNullAssertion: <explanation>
					process.env.UPLOAD_KEY!
				}`,
			},
			body: JSON.stringify({
				uploadId,
			}),
		},
	);

	if (!resp.ok) {
		const error = await resp.text();
		const errorMsg = `Unable to abort multipart upload. ${error}`;
		handleBucketError(resp, msg ? `${msg} ${errorMsg}` : errorMsg);
	}
};

const initiateMultipartUpload = async (bucketPath: string): Promise<string> => {
	const resp = await fetch(
		`https://upload.fontsource.org/multipart/${bucketPath}?action=mpu-create`,
		{
			method: 'POST',
			headers: {
				Authorization: `Bearer ${
					// biome-ignore lint/style/noNonNullAssertion: <explanation>
					process.env.UPLOAD_KEY!
				}`,
			},
		},
	);

	if (!resp.ok) {
		const error = await resp.text();
		handleBucketError(resp, `Unable to initiate multipart upload. ${error}`);
	}

	const data = await resp.json();
	if (!data.uploadId) {
		throw new StatusError(
			500,
			`Internal Server Error. Upload ID is missing. ${JSON.stringify(data)}`,
		);
	}
	return data.uploadId;
};

const uploadPart = async (
	bucketPath: string,
	uploadId: string,
	partNumber: number,
	partData: Uint8Array | ArrayBuffer,
) => {
	keepAwake(SLEEP_MINUTES);
	const formData = new FormData();
	formData.append('partNumber', String(partNumber));
	formData.append('uploadId', uploadId);

	const blob = new Blob([partData]);
	const filename = `${bucketPath}-${partNumber}`;
	formData.append('file', blob, filename);

	const resp = await fetch(
		`https://upload.fontsource.org/multipart/${bucketPath}?action=mpu-uploadpart`,
		{
			method: 'PUT',
			headers: {
				Authorization: `Bearer ${
					// biome-ignore lint/style/noNonNullAssertion: <explanation>
					process.env.UPLOAD_KEY!
				}`,
			},
			body: formData,
		},
	);

	if (!resp.ok) {
		const error = await resp.text();
		const msg = `Unable to upload part. ${error}`;

		await abortMultiPartUpload(bucketPath, uploadId, msg);
		handleBucketError(resp, msg);
	}

	const data = await resp.json();
	if (!data.etag) {
		throw new StatusError(
			500,
			`Internal Server Error. ETag is missing. ${JSON.stringify(data)}`,
		);
	}
	return data.etag;
};

const completeMultipartUpload = async (
	bucketPath: string,
	uploadId: string,
	parts: R2UploadedPart[],
) => {
	const resp = await fetch(
		`https://upload.fontsource.org/multipart/${bucketPath}?action=mpu-complete`,
		{
			method: 'POST',
			headers: {
				Authorization: `Bearer ${
					// biome-ignore lint/style/noNonNullAssertion: <explanation>
					process.env.UPLOAD_KEY!
				}`,
				'Content-Type': 'application/json',
			},
			body: JSON.stringify({
				uploadId,
				parts,
			}),
		},
	);

	if (!resp.ok) {
		const error = await resp.text();
		const msg = `Unable to complete multipart upload. ${error}`;

		await abortMultiPartUpload(bucketPath, uploadId, msg);
		handleBucketError(resp, msg);
	}
};

export const putBucket = async (
	bucketPath: string,
	body: Uint8Array | ArrayBuffer,
) => {
	// We only use multipart uploads for files larger than 80MB since Cloudflare
	// limits the maximum request size to 100MB
	const partSize = 80 * 1024 * 1024;
	if (body.byteLength < partSize) {
		keepAwake(SLEEP_MINUTES);
		const resp = await fetch(
			`https://upload.fontsource.org/put/${bucketPath}`,
			{
				method: 'PUT',
				headers: {
					Authorization: `Bearer ${
						// biome-ignore lint/style/noNonNullAssertion: <explanation>
						process.env.UPLOAD_KEY!
					}`,
				},
				body,
			},
		);

		if (!resp.ok) {
			const error = await resp.text();
			handleBucketError(resp, `Unable to upload file ${bucketPath}. ${error}`);
		}

		return;
	}

	info(`Uploading ${bucketPath} in parts with size ${body.byteLength}`);

	const uploadId = await initiateMultipartUpload(bucketPath);

	const parts: R2UploadedPart[] = [];
	let offset = 0;
	let partNumber = 1;

	// Upload buffers in parts
	while (offset < body.byteLength) {
		const end = Math.min(offset + partSize, body.byteLength);
		const partData = body.slice(offset, end);
		info(`Uploading part ${partNumber} with size ${partData.byteLength}`);

		const etag = await uploadPart(bucketPath, uploadId, partNumber, partData);

		parts.push({
			etag,
			partNumber,
		});

		offset = end;
		partNumber++;
	}

	// Complete multipart upload
	await completeMultipartUpload(bucketPath, uploadId, parts);
};

export const getBucket = async (bucketPath: string) => {
	keepAwake(SLEEP_MINUTES);

	const resp = await fetch(`https://upload.fontsource.org/get/${bucketPath}`, {
		headers: {
			Authorization: `Bearer ${
				// biome-ignore lint/style/noNonNullAssertion: <explanation>
				process.env.UPLOAD_KEY!
			}`,
		},
	});
	if (!resp.ok) {
		handleBucketError(resp, `Unable to fetch ${bucketPath}.`);
	}

	return resp;
};
