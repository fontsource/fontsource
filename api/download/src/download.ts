import { info } from 'diary';
import { StatusError } from 'itty-router';
import PQueue from 'p-queue';
// @ts-expect-error - no types
import woff2ttf from 'woff2sfnt-sfnt2woff';

import {
	bucketPath,
	bucketPathVariable,
	getBucket,
	listBucket,
	putBucket,
} from './bucket';
import {
	generateManifest,
	type Manifest,
	type ManifestVariable,
} from './manifest';
import { keepAwake, SLEEP_MINUTES } from './sleep';
import { type IDResponse } from './types';

export const downloadFile = async (manifest: Manifest) => {
	const { id, subset, weight, style, extension, version, url } = manifest;

	const res = await fetch(url);
	info(`Downloading ${url}`);

	if (!res.ok) {
		throw new StatusError(500, `Could not fetch ${url}`);
	}

	const buffer = await res.arrayBuffer();

	// Add to bucket
	await putBucket(bucketPath(manifest), buffer);

	// If woff, decompress and add to ttf folder
	if (extension === 'woff') {
		let ttfBuffer;
		try {
			ttfBuffer = await woff2ttf.toSfnt(new Uint8Array(buffer));
		} catch (error) {
			throw new StatusError(
				500,
				`Could not convert woff to ttf ${String(error)}`,
			);
		}
		if (!ttfBuffer) throw new StatusError(500, 'Could not convert woff to ttf');

		// Add to bucket
		await putBucket(
			bucketPath({
				id,
				subset,
				weight,
				style,
				extension: 'ttf',
				version,
			}),
			ttfBuffer,
		);
	}
};

export const downloadVariableFile = async (manifest: ManifestVariable) => {
	const { url } = manifest;
	const res = await fetch(url);

	if (!res.ok) {
		throw new StatusError(500, `Could not fetch ${url}`);
	}

	const buffer = await res.arrayBuffer();

	// Add to bucket
	await putBucket(bucketPathVariable(manifest), buffer);
};

export const downloadManifest = async (manifest: Manifest[]) => {
	// Create a queue
	const queue = new PQueue({ concurrency: 24 });
	let hasError: Error | undefined;

	// Download all files
	for (const file of manifest) {
		// eslint-disable-next-line @typescript-eslint/promise-function-async
		queue
			.add(async () => {
				await downloadFile(file);
			})
			// eslint-disable-next-line no-loop-func
			.catch((error) => {
				queue.pause();
				queue.clear();
				hasError = error;
			});
	}

	// Wait for all files to be downloaded
	await queue.onIdle();
	if (hasError) throw hasError;
};

export const generateZip = async (
	id: string,
	version: string,
	metadata: IDResponse,
) => {
	// Check if zip file already exists
	const zipFile = await listBucket(`${id}@${version}/download.zip`);
	if (zipFile.objects.length > 0) return;

	info(`Generating zip file for ${id}@${version}`);

	const fullManifest = generateManifest(`${id}@${version}`, metadata);
	// For every woff file, generate an equivalent manifest entry for ttf
	for (const file of fullManifest) {
		if (file.extension === 'woff') {
			fullManifest.push({
				...file,
				extension: 'ttf',
			});
		}
	}

	// Generate zip file of all fonts
	const zip = [];
	const zipQueue = new PQueue({ concurrency: 32 });

	// Download all files
	for (const file of fullManifest) {
		// eslint-disable-next-line @typescript-eslint/promise-function-async
		zipQueue
			.add(async () => {
				keepAwake(SLEEP_MINUTES);
				const item = await getBucket(bucketPath(file));
				if (!item) {
					throw new StatusError(500, `Could not find ${bucketPath(file)}`);
				}

				const buffer = await item.arrayBuffer();
				info(`Deflating ${bucketPath(file)}`);
				const deflatedBuffer = Bun.deflateSync(new Uint8Array(buffer));

				// Add to zip
				if (file.extension === 'woff2' || file.extension === 'woff') {
					zip.push({
						filename: `webfonts/${file.id}-${file.subset}-${file.weight}-${file.style}.${file.extension}`,
						content: deflatedBuffer,
					});
				} else if (file.extension === 'ttf') {
					zip.push({
						filename: `ttf/${file.id}-${file.subset}-${file.weight}-${file.style}.${file.extension}`,
						content: deflatedBuffer,
					});
				} else {
					throw new StatusError(
						500,
						`Invalid file extension ${file.extension}`,
					);
				}
			})
			.catch((error) => {
				zipQueue.pause();
				zipQueue.clear();
				throw error;
			});
	}

	// Add LICENSE file
	const license = await fetch(
		`https://cdn.jsdelivr.net/npm/@fontsource/${id}@${version}/LICENSE`,
	);
	if (!license.ok) {
		throw new StatusError(500, 'Could not find LICENSE file');
	}

	const licenseBuffer = await license.arrayBuffer();
	const deflatedLicenseBuffer = Bun.deflateSync(new Uint8Array(licenseBuffer));
	zip.push({
		filename: 'LICENSE',
		content: deflatedLicenseBuffer,
	});

	// Combine zip file
	const zipLength = zip.reduce((total, file) => total + file.content.length, 0);
	const zipBuffer = new Uint8Array(zipLength);

	let offset = 0;
	for (const file of zip) {
		zipBuffer.set(file.content, offset);
		offset += file.content.length;
	}

	// Add to bucket
	info(`Uploading zip file for ${id}@${version}`);
	await putBucket(`${id}@${version}/download.zip`, zipBuffer);
	info(`Successfully generated zip file for ${id}@${version}`);
};
