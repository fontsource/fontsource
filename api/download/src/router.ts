import { getMetadata } from 'common-api/util';
import {
	error,
	type IRequestStrict,
	Router,
	text,
	withParams,
} from 'itty-router';

import { type CFRouterContext } from './types';
import {
	bucketPath,
	downloadFile,
	downloadManifest,
	generateZip,
} from './update';
import { generateManifest, generateManifestItem } from './util';

interface DownloadRequest extends IRequestStrict {
	id: string;
	file: string;
}

const router = Router<DownloadRequest, CFRouterContext>();

router.post('/v1/:id', withParams, async (request, env, _ctx) => {
	const { id } = request;
	const fontId = id.split('@')[0];

	const metadata = await getMetadata(fontId, request.clone(), env);
	if (!metadata) {
		return error(404, 'Not Found. Font does not exist.');
	}

	const baseManifest = await generateManifest(id, metadata);

	// Search for a list of existing files and prune out those that already exist
	const existingFiles = await env.BUCKET.list({
		prefix: `${id}/`,
	});

	const manifest = baseManifest.filter((file) => {
		const existingFile = existingFiles.objects.find((existingFile) => {
			return existingFile.key === bucketPath(file);
		});

		return !existingFile;
	});

	await downloadManifest(manifest, env);
	await generateZip(manifest[0].id, manifest[0].version, request, env);

	return text('Success.');
});

router.post('/v1/:id/:file', withParams, async (request, env, _ctx) => {
	const { id, file } = request;
	const fontId = id.split('@')[0];

	const metadata = await getMetadata(fontId, request.clone(), env);
	if (!metadata) {
		return error(404, 'Not Found. Font does not exist.');
	}

	const manifestItem = await generateManifestItem(id, file, metadata);
	await downloadFile(manifestItem, env);

	return text('Success.');
});

// 404 for everything else
router.all('*', () =>
	error(
		404,
		'Not Found. Please refer to the Fontsource API documentation: https://fontsource.org/docs/api',
	),
);

export default router;
