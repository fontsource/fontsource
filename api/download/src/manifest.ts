import { info } from 'diary';
import { StatusError } from 'itty-router';

import { bucketPath, listBucket } from './bucket';
import { type IDResponse } from './types';
import { splitTag } from './util';

export interface Manifest {
	id: string;
	subset: string;
	weight: number;
	style: string;
	extension: string;
	version: string;
	url: string;
}

export interface ManifestVariable {
	id: string;
	subset: string;
	axes: string;
	style: string;
	version: string;
	url: string;
}

type ManifestGenerator = Omit<
	IDResponse,
	| 'variants'
	| 'defSubset'
	| 'license'
	| 'type'
	| 'family'
	| 'lastModified'
	| 'category'
>;

export const generateManifestItem = (
	tag: string,
	file: string,
	metadata: ManifestGenerator,
): Manifest => {
	const { id, version } = splitTag(tag);
	let [filename, extension] = file.split('.');
	const filenameArr = filename.split('-');
	const style = filenameArr.pop();
	const weight = filenameArr.pop();
	const subset = filenameArr.join('-');
	if (!subset || !weight || !style) {
		throw new StatusError(400, 'Bad Request. Invalid filename.');
	}

	if (id !== metadata.id) {
		throw new StatusError(400, 'Bad Request. Invalid ID.');
	}

	// If the extension is ttf, change it to woff since jsdelivr doesn't store tff
	// and we convert it from woff to ttf on the fly
	if (extension === 'ttf') {
		extension = 'woff';
	}

	return {
		id,
		subset,
		weight: Number(weight),
		style,
		extension,
		version,
		url: `https://cdn.jsdelivr.net/npm/@fontsource/${id}@${version}/files/${id}-${file.replace(
			'.ttf',
			'.woff',
		)}`,
	};
};

export const generateVariableManifestItem = (
	tag: string,
	file: string,
	metadata: ManifestGenerator,
): ManifestVariable => {
	const { id, version } = splitTag(tag);
	const [filename, extension] = file.split('.');
	const filenameArr = filename.split('-');
	const style = filenameArr.pop();
	const axes = filenameArr.pop();
	const subset = filenameArr.join('-');
	if (!subset || !axes || !style) {
		throw new StatusError(400, 'Bad Request. Invalid filename.');
	}
	if (extension !== 'woff2') {
		throw new StatusError(400, 'Bad Request. Invalid extension.');
	}

	if (id !== metadata.id) {
		throw new StatusError(
			422,
			'Unprocessable Entity.. Provided tag does not match metadata ID.',
		);
	}

	return {
		id,
		subset,
		axes,
		style,
		version,
		url: `https://cdn.jsdelivr.net/npm/@fontsource-variable/${id}@${version}/files/${id}-${file}`,
	};
};

export const generateManifest = (
	tag: string,
	metadata: ManifestGenerator,
): Manifest[] => {
	const { id, version } = splitTag(tag);

	if (id !== metadata.id) {
		throw new StatusError(
			422,
			'Unprocessable Entity.. Provided tag does not match metadata ID.',
		);
	}

	// Generate manifest
	const manifest: Manifest[] = [];

	// Unicode range is needed to include numbered subset files, but it may include duplicates
	const subsetKeys = [
		...new Set([
			...Object.keys(metadata.unicodeRange).map(
				(key) => key.replace('[', '').replace(']', ''), // Files should be stored without brackets
			),
			...metadata.subsets,
		]),
	];

	for (const subset of subsetKeys) {
		for (const weight of metadata.weights) {
			for (const style of metadata.styles) {
				for (const extension of ['woff2', 'woff']) {
					manifest.push({
						id,
						subset,
						weight,
						style,
						extension,
						version,
						// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
						url: `https://cdn.jsdelivr.net/npm/@fontsource/${id}@${version}/files/${metadata.id}-${subset}-${weight}-${style}.${extension}`,
					});
				}
			}
		}
	}

	return manifest;
};

// Search for a list of existing files and prune out those that already exist
export const pruneManifest = async (
	id: string,
	version: string,
	baseManifest: Manifest[],
	isVariable?: boolean,
) => {
	const prefix = isVariable
		? `${id}@${version}/variable/`
		: `${id}@${version}/`;
	const existingFiles = await listBucket(prefix);

	// Filter out existing files
	const manifest = [];
	const existingSet = new Set(existingFiles.objects);
	for (const file of baseManifest) {
		if (!existingSet.has(bucketPath(file))) {
			manifest.push(file);
		}
	}

	info(
		`Found ${manifest.length} new files to download out of ${baseManifest.length} from original manifest.`,
	);
	return manifest;
};
