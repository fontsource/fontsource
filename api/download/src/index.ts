import { enable, error as errorDiary, info } from 'diary';
import { error as errorRes, json, type StatusError } from 'itty-router';

import router from './router';

enable('*');

export default {
	port: 3001,
	fetch: async (req: Request) => {
		return await router
			.handle(req)
			.then((res: Response) => {
				const path = new URL(req.url).pathname;
				info(`${req.method} ${path} ${res.status}`);
				return json(res);
			})
			.catch((error_: StatusError) => {
				const path = new URL(req.url).pathname;
				if (error_.status >= 500) {
					errorDiary(
						`${req.method} ${path} ${error_.status} ${error_.message}`,
					);
				} else {
					info(`${req.method} ${path} ${error_.status} ${error_.message}`);
				}

				return errorRes(error_);
			});
	},
};
