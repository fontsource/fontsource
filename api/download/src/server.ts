import { enable, error as errorDiary, info } from 'diary';
import { error as errorRes, json, type StatusError } from 'itty-router';

import router from './router';

// Enable logging
enable('*');

// Have a sleep timer that kills the worker after 5 minutes
const SLEEP_MINUTES = 2;
let sleepTimeout: NodeJS.Timeout;
const keepAwake = (minutes: number) => {
	if (sleepTimeout) {
		clearTimeout(sleepTimeout);
	}

	// eslint-disable-next-line unicorn/no-process-exit
	sleepTimeout = setTimeout(() => process.exit(0), 1000 * 60 * minutes);
};

// Trigger the timeout on first load
keepAwake(SLEEP_MINUTES);

export default {
	port: process.env.PORT ?? 3001,
	fetch: async (req: Request) => {
		return await router
			.handle(req)
			.then((res: Response) => {
				const path = new URL(req.url).pathname;
				info(`${req.method} ${path} ${res.status}`);
				keepAwake(SLEEP_MINUTES);
				return json(res);
			})
			.catch((error_: StatusError) => {
				const path = new URL(req.url).pathname;
				if (error_.status >= 500) {
					errorDiary(
						`${req.method} ${path} ${error_.status} ${error_.message}`,
					);
				} else {
					info(`${req.method} ${path} ${error_.status} ${error_.message}`);
				}

				keepAwake(SLEEP_MINUTES);
				return errorRes(error_);
			});
	},
};
