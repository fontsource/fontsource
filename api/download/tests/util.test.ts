import { expect, it } from 'vitest';

import { generateManifest } from '../src/util';

const describe = setupMiniflareIsolatedStorage();

describe('utils', () => {
	it('should generate valid manifest', async () => {
		const metadata = {
			weights: [400, 700],
			styles: ['normal', 'italic'],
			subsets: ['latin'],
			variable: false,
			unicodeRange: {
				latin: 'U+0000-00FF',
			},
		};

		const manifest = generateManifest('roboto@latest', metadata);
	});
});
