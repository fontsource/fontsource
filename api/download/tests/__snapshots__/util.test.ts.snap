// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > should generate valid manifest 1`] = `
[
  {
    "extension": "woff2",
    "id": "roboto",
    "style": "normal",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-400-normal.woff2",
    "variable": false,
    "version": "5.0.0",
    "weight": 400,
  },
  {
    "extension": "woff",
    "id": "roboto",
    "style": "normal",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-400-normal.woff",
    "variable": false,
    "version": "5.0.0",
    "weight": 400,
  },
  {
    "extension": "woff2",
    "id": "roboto",
    "style": "italic",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-400-italic.woff2",
    "variable": false,
    "version": "5.0.0",
    "weight": 400,
  },
  {
    "extension": "woff",
    "id": "roboto",
    "style": "italic",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-400-italic.woff",
    "variable": false,
    "version": "5.0.0",
    "weight": 400,
  },
  {
    "extension": "woff2",
    "id": "roboto",
    "style": "normal",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-700-normal.woff2",
    "variable": false,
    "version": "5.0.0",
    "weight": 700,
  },
  {
    "extension": "woff",
    "id": "roboto",
    "style": "normal",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-700-normal.woff",
    "variable": false,
    "version": "5.0.0",
    "weight": 700,
  },
  {
    "extension": "woff2",
    "id": "roboto",
    "style": "italic",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-700-italic.woff2",
    "variable": false,
    "version": "5.0.0",
    "weight": 700,
  },
  {
    "extension": "woff",
    "id": "roboto",
    "style": "italic",
    "subset": "latin",
    "url": "https://cdn.jsdelivr.net/npm/@fontsource/roboto@5.0.0/files/roboto-latin-700-italic.woff",
    "variable": false,
    "version": "5.0.0",
    "weight": 700,
  },
]
`;
